src/sentry/search/snuba/backend.py

```diff
@@ -624,7 +489,24 @@ def build_results(
                 if not snuba_groups:
                     break
 
-                if candidate_ids:
+                if skip_postgres_search:
```

Comment:
This approach will fail at line 448, since now group_queryset can be None, but I can still get there if the request is being paginated. count_hits can be True and cursor can be not None even when we skip the postgres query.

So I have two main concerns with this approach. One is about complexity and the other one is about correctness.
Correctness: hits is not total. The snuba query is done in chunks. I think the easiest way to get populate hits is to perform a separate Snuba query(which would just return the count). This is not too different than the one we do at line 432 to return the count. Of course we are trading some savings in not querying postgres with an additional snuba query. Then when we figure out how expensive this is we can decide if we need sampling.

Complexity: as mentioned before in the parent PR, it seems the whole process to make a query with postgres is quite different than the one on snuba only, the difference is not only simply concerned with the parameters we pass to Snuba.

To abstract this:
this is the process for a postgres + snuba query:
- query postgres to find candidate groups
- calculate hits by making a sampled query on snuba and statistically approximate the real value through the postgres results
- break the snuba query into time based chunks, run the snuba query
- post filter the result if needed

this is the process for a snuba only query:
- calculate hits by making an aggregate snuba query. No postgres. We can optimize this later either with sampling or by approximate count, etc.
- break the snuba query into time based chunks, run the snuba query

Fitting the two query paths in the same function increase the complexity of the query method by a factor higher than two making it very hard to review, test and maintain because of the increased set of code paths. If you fit the logic of the two separate methods into this function and use `if skip_postgres_search:` to distinguish, you are not only duplicating the complexity, you are increasing the potential number of code path to think about by 2 ^ number_of_if_statements_above. This is because it is not immediately evident which code paths are not possible.
You can make this a lot easier be pushing more logic into the subclasses as I suggested already in the previous review and in the review of the parent PR. 
You can do this in two approaches:
A) Just cut paste all the logic between line 355 and 550 and move it in a method that is implemented in bpth subclasses. This will create initially a lot of copy/paste but it will make this method much easier since now you have much fewer code paths to understand to read this code: one with postgres and one without. Once you have done this you can reduce copy paste by copying some smaller pieces of codes in their own methods and reuse them. At that point this last step is trivial.
B) introduce two methods you implement in the subclasses, one per phase in the query: `calculate_candidates`, `calculate_hits`, `query_snuba`. So the parent class at least does not know how the hits are calculated, or if post filtering needs to be done, thus simplifying the code path of the parent class.

I prefer A (this is what we discussed last week in the meeting), but I think B ma end up being acceptable if A is too time consuming to implement.